pkg/util/sqlkiller/sqlkiller.go

```diff
@@ -41,11 +41,47 @@ const (
 type SQLKiller struct {
 	Signal killSignal
 	ConnID uint64
+	Finish func()
+	// InWriteResultSet is used to mark whether the query is calling clientConn.writeResultSet().
+	// If the query is in writeResultSet, and Finish() can acquire rs.finishLock, we can assume the query is waiting for network IO.
+	InWriteResultSet atomic.Bool
 }
 
 // SendKillSignal sends a kill signal to the query.
 func (killer *SQLKiller) SendKillSignal(reason killSignal) {
-	atomic.CompareAndSwapUint32(&killer.Signal, 0, reason)
+	if atomic.CompareAndSwapUint32(&killer.Signal, 0, reason) {
+		status := atomic.LoadUint32(&killer.Signal)
+		err := killer.getKillError(status)
+		logutil.BgLogger().Warn("kill query started", zap.Uint64("conn", killer.ConnID), zap.String("reason", err.Error()))
+	}
+}
+
+// GetKillSignal gets the kill signal.
+func (killer *SQLKiller) GetKillSignal() killSignal {
+	return atomic.LoadUint32(&killer.Signal)
+}
+
+// getKillError gets the error according to the kill signal.
+func (killer *SQLKiller) getKillError(status killSignal) error {
+	switch status {
+	case QueryInterrupted:
+		return exeerrors.ErrQueryInterrupted.GenWithStackByArgs()
+	case MaxExecTimeExceeded:
+		return exeerrors.ErrMaxExecTimeExceeded.GenWithStackByArgs()
+	case QueryMemoryExceeded:
+		return exeerrors.ErrMemoryExceedForQuery.GenWithStackByArgs(killer.ConnID)
+	case ServerMemoryExceeded:
+		return exeerrors.ErrMemoryExceedForInstance.GenWithStackByArgs(killer.ConnID)
+	}
+	return nil
+}
+
+// FinishResultSet is used to finish the result set.
+// If the cancel signal is received and SQL is waiting for network IO, resource released can be performed first.
```

Comment:
// FinishResultSet is used to close the result set.
// If a kill signal is sent but the SQL query is stuck in the network stack while writing packets to the client,
// encountering some bugs that cause it to hang, or failing to detect the kill signal, we can call Finish to release resources used during the SQL execution process.